### PR TITLE
Convert netty 4.0 test to java

### PR DESCRIPTION
**Pull request by heyams**
part of #7195

### CODE REVIEW

File: instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/ChannelPipelineTest.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+    // http and instrumentation handlers were removed; noop handler was added
+    assertEquals(1, channelPipeline.toMap().size());
+    assertEquals(anotherNoopHandler, channelPipeline.first());
+  }
+
+  // regression test for
+  // https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/4056
+  @Test
+  void shouldAddAfterAndRemoveLastHandler() {
```
Comment (heyams) [inline instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/ChannelPipelineTest.java:105]: i modified this test based on current implementation. please let me know if this is not the case, might be regression? 
i might need to update netty 4.0 instrumentation for this. but i need some feedback here.

File: instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/ChannelPipelineTest.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    NoopChannelHandler noopHandler = new NoopChannelHandler();
+    channelPipeline.addAfter("http", "noop", noopHandler);
+
+    // http -> instrumentation -> noop
+    // instrumentation handler is between with http and noop
+    assertEquals(3, channelPipeline.toMap().size());
+    assertEquals(httpHandler, channelPipeline.first());
+    assertEquals(noopHandler, channelPipeline.last());
+
+    // removeLast will remove everything after http
```
Comment (heyams) [inline instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/ChannelPipelineTest.java:134]: removeLast will remove everything after http.

File: instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/ChannelPipelineTest.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    EmbeddedChannel channel = new EmbeddedChannel(new NoopChannelHandler());
+    ChannelPipeline channelPipeline = channel.pipeline();
+    HttpClientCodec httpHandler = new HttpClientCodec();
+
+    // default: head -> tail
+    channelPipeline.removeFirst();
+    channelPipeline.removeLast();
+
+    // start with no handlers initially
+    assertEquals(0, channelPipeline.toMap().size());
```
Comment (heyams) [inline instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/ChannelPipelineTest.java:115]: even though size is 0, but channelPipeline.first() and channelPipeline.last() are not null.

**Comment by heyams** [comment]
need some help here: 

image
 
Netty40ServerTest is green locally, but CI complained about KtorHttpServerTest: 

KtorHttpServerTest > requestWithException() STANDARD_OUT
2023-12-08T23:21:58.5351175Z     23:21:55.534 [eventLoopGroupProxy-4-4 @call-handler#1829] ERROR ktor.application - Unhandled: GET - /exception
2023-12-08T23:21:58.5352525Z     java.lang.Exception: controller exception
2023-12-08T23:21:58.5353395Z     	at _COROUTINE._BOUNDARY._(CoroutineDebugging.kt:46)
2023-12-08T23:21:58.5355015Z     	at io.opentelemetry.instrumentation.ktor.v2_0.server.AbstractKtorHttpServerTest.controller(AbstractKtorHttpServerTest.kt:109)
2023-12-08T23:21:58.5357449Z     	at io.opentelemetry.instrumentation.ktor.v2_0.server.AbstractKtorHttpServerTest$setupServer$1$1$4.invokeSuspend(AbstractKtorHttpServerTest.kt:59)
2023-12-08T23:21:58.5359435Z     	at io.ktor.server.routing.Route$buildPipeline$1$1.invokeSuspend(Route.kt:116)
2023-12-08T23:21:58.5360836Z     	at io.ktor.util.pipeline.PipelineKt$execute$2.invokeSuspend(Pipeline.kt:478)

is this sporadic?

**Comment by heyams** [comment]
**rerun failed jobs** button is nowhere to be found. do i need additional permission to get access to it?

**Comment by laurit** [comment]
> **rerun failed jobs** button is nowhere to be found. do i need additional permission to get access to it?

I believe so, you can push an empty commit to trigger checks. Closing and reopening the pr should also work.

**Comment by jaydeluca** [comment]
@heyams I suspect the test failures might not be the ktor tests you referenced. Your screenshot shows passing tests when running the `Netty40ServerTest`, but have you tried running the Client tests? I ran your tests locally and the `Netty40ClientTest` hangs forever

image

There are also different variations of the tests

image

When trying to run some of those (`:instrumentation:netty:netty-4.0:javaagent:testConnectionSpan`)  I am seeing some failures there as well:

image

**Comment by heyams** [comment]
https://github.com/open-telemetry/opentelemetry-java-instrumentation/actions/runs/7200754364/job/19615382707?pr=9976

stuck with this error. can someone help me take a look at it? 

KtorHttpServerTest is green locally.. but CI continued to fail.

**Comment by jaydeluca** [comment]
@heyams great progress! I am able to see that all tests pass now. The issue now seems to be some failing tests when running with `-PtestLatestDeps=true` 

examples: `./gradlew :instrumentation:netty:netty-4.0:javaagent:test -PtestLatestDeps=true`

image


`./gradlew :instrumentation:netty:netty-4.0:javaagent:testStableSemconv -PtestLatestDeps=true` also has failures

**Comment by heyams** [comment]
I'm not sure i can spot the cause of remaining failures. Huge log file and looking at raw log, didn't say which test failed exactly. Test Summary wasn't helpful either. Appreciate any tips or tricks...

**Comment by jaydeluca** [comment]
@heyams it seems there are still a few failures. I am not able to reproduce the failures locally but I downloaded the logs and found an example of one of the remaining failures:

```
2023-12-14T01:06:49.6340342Z > Task :instrumentation:netty:netty-4.0:javaagent:testConnectionSpan
2023-12-14T01:06:49.6344250Z     [otel.javaagent 2023-12-14 01:06:44:869 +0000] [otel-javaagent-transform-safe-logger] DEBUG io.opentelemetry.javaagent.tooling.AgentInstaller$TransformLoggingListener - Transformed io.opentelemetry.instrumentation.testing.util.ContextStorageCloser$$Lambda$1037 -- sun.misc.Launcher$AppClassLoader@73d16e93
2023-12-14T01:06:49.6347068Z 
2023-12-14T01:06:49.6347506Z Netty40ConnectionSpanTest > failedRequest() FAILED
2023-12-14T01:06:49.6348787Z     org.opentest4j.AssertionFailedError: [STRING attribute 'network.type'] 
2023-12-14T01:06:49.6349747Z     expected: "ipv4"
2023-12-14T01:06:49.6350255Z      but was: null
```

In some of the groovy tests the assertions handle an `or` condition ([example](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/600b14db8c53bf9c8ae1f3eae137918a6a4bfaae/instrumentation/netty/netty-4.0/javaagent/src/test/groovy/Netty40ConnectionSpanTest.groovy#L159))

So you need to account for that in places like [here](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/9976/files#diff-5304956b5ed5308500dd4977c79306f7aa8448769240165790187195530f38a6R135). 

here [is an example](https://github.com/jaydeluca/opentelemetry-java-instrumentation/blob/main/instrumentation/camel-2.20/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/apachecamel/aws/AwsSpanAssertions.java#L74-L79) for reference of how you can solve for that.

I'm not sure how to reproduce this failure locally unfortunately

File: instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/client/Netty40ClientSslTest.java [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+import spock.lang.Shared;
+
+class Netty40ClientSslTest {
+
+  @RegisterExtension
+  static final InstrumentationExtension testing = AgentInstrumentationExtension.create();
+
+  @Shared static HttpClientTestServer server;
```
Comment (jaydeluca) [inline instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/client/Netty40ClientSslTest.java:57]: `@Shared` is a groovy package so we don't need that

File: instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/ChannelFutureTest.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+import io.netty.util.concurrent.ProgressiveFuture;
+import io.opentelemetry.instrumentation.testing.junit.AgentInstrumentationExtension;
+import io.opentelemetry.instrumentation.testing.junit.InstrumentationExtension;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.jetbrains.annotations.NotNull;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+public class ChannelFutureTest {
```
Comment (jaydeluca) [inline instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/ChannelFutureTest.java:25]: Several of these Test classes are public but could be package private instead

**Comment by heyams** [comment]
@jaydeluca it's finally green. your comments helped me a lot. thank you

File: instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/server/Netty40ServerTest.java [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+import io.opentelemetry.instrumentation.testing.junit.http.HttpServerTestOptions;
+import io.opentelemetry.instrumentation.testing.junit.http.ServerEndpoint;
+import io.opentelemetry.semconv.SemanticAttributes;
+import java.net.URI;
+import java.util.HashSet;
+import java.util.Set;
+import org.jetbrains.annotations.NotNull;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+public class Netty40ServerTest extends AbstractHttpServerTest<EventLoopGroup> {
```
Comment (jeanbisutti) [inline instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/server/Netty40ServerTest.java:58]:
```suggestion
class Netty40ServerTest extends AbstractHttpServerTest<EventLoopGroup> {
```

File: instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/client/Netty40ClientTest.java [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import org.jetbrains.annotations.NotNull;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+public class Netty40ClientTest extends AbstractHttpClientTest<DefaultFullHttpRequest> {
```
Comment (jeanbisutti) [inline instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/client/Netty40ClientTest.java:41]:
```suggestion
class Netty40ClientTest extends AbstractHttpClientTest<DefaultFullHttpRequest> {
```

File: instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/client/ClientHandler.java [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+ */
+
+package io.opentelemetry.javaagent.instrumentation.netty.v4_0.client;
+
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.handler.codec.http.HttpResponse;
+import java.util.concurrent.CompletableFuture;
+
+public class ClientHandler extends ChannelInboundHandlerAdapter {
```
Comment (jeanbisutti) [inline instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/client/ClientHandler.java:13]:
```suggestion
class ClientHandler extends ChannelInboundHandlerAdapter {
```

File: instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/ChannelPipelineTest.java [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+public class ChannelPipelineTest {
+
+  @RegisterExtension
+  static final InstrumentationExtension testing = AgentInstrumentationExtension.create();
+
+  @ParameterizedTest
+  @CsvSource({"by instance", "by class", "by name", "first"})
+  void testRemoveOurHandler(String testName) throws Exception {
+    EmbeddedChannel channel = new EmbeddedChannel(new NoopChannelHandler());
+    ChannelPipeline channelPipeline = channel.pipeline();
```
[the remark this answers is not in the document]
Comment (laurit) [reply instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/ChannelPipelineTest.java:34]: you could get around this by creating the instance with reflection or by placing a helper class in the same package as `DefaultChannelPipeline` and call the constructor using that class

File: instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/ChannelPipelineTest.java [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+    if ("by instance".equals(desc)) {
+      channelPipeline.replace(noopHandler, "http", httpHandler);
+    } else if ("by class".equals(desc)) {
+      channelPipeline.replace(noopHandler.getClass(), "http", httpHandler);
+    } else if ("by name".equals(desc)) {
+      channelPipeline.replace("test", "http", httpHandler);
+    }
+
+    // noop handler was removed; http and instrumentation handlers were added
+    assertEquals(2, channelPipeline.toMap().size());
```
Comment (laurit) [inline instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/ChannelPipelineTest.java:104]: https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/10410 removes `HttpClientTracingHandler` from the map so the size here will be `1` after you rebase your pr

File: instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/ChannelPipelineTest.java [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+    assertEquals(httpHandler, channelPipeline.first());
+    assertEquals("HttpClientTracingHandler", channelPipeline.last().getClass().getSimpleName());
+
+    NoopChannelHandler anotherNoopHandler = new NoopChannelHandler();
+    channelPipeline.replace("http", "test", anotherNoopHandler);
```
Comment (laurit) [inline instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/ChannelPipelineTest.java:109]: in groovy test here different replace method is used depending on the parameter

File: instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/ChannelPipelineTest.java [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+    assertEquals(channelPipeline.first(), httpHandler);
+    assertEquals(channelPipeline.last().getClass().getSimpleName(), "HttpClientTracingHandler");
+
+    NoopChannelHandler noopHandler = new NoopChannelHandler();
+    channelPipeline.addAfter("http", "noop", noopHandler);
+
+    assertEquals(3, channelPipeline.toMap().size());
+    assertEquals(channelPipeline.first(), httpHandler);
+    assertEquals(channelPipeline.last(), noopHandler);
+
```
Comment (laurit) [inline instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/ChannelPipelineTest.java:144]: as far as I can tell there was a bit more code here in the groovy test

File: instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/client/Netty40ClientSslTest.java [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+class Netty40ClientSslTest {
+
+  @RegisterExtension
+  static final InstrumentationExtension testing = AgentInstrumentationExtension.create();
+
+  private static HttpClientTestServer server;
+  private static EventLoopGroup eventLoopGroup;
+
+  @BeforeAll
+  static void setupSpec() {
```
Comment (laurit) [inline instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/client/Netty40ClientSslTest.java:60]: `setupSpec` and `cleanupSpec` are spock method names, in java tests we usually use `setup`/`setUp` and `cleanup`/`cleanUp`/`tearDown`

File: instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/client/Netty40ClientSslTest.java [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+        "parent",
+        () -> {
+          try {
+            channel.set(bootstrap.connect(uri.getHost(), uri.getPort()).sync().channel());
+            CompletableFuture<Integer> result = new CompletableFuture<>();
+            channel.get().pipeline().addLast(new ClientHandler(result));
+            channel.get().writeAndFlush(request).get(10, TimeUnit.SECONDS);
+            result.get(10, TimeUnit.SECONDS);
+          } catch (Exception ex) {
+            throw ex;
```
Comment (laurit) [inline instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/client/Netty40ClientSslTest.java:164]: just catching and rethrowning the exception isn't useful

File: instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/client/Netty40ClientSslTest.java [hunk 15]
```diff
@@ -0,0 +1,244 @@
...
+                },
+                span -> {
+                  span.hasName("GET").hasKind(CLIENT).hasParent(trace.getSpan(0));
+                },
+                span -> {
+                  span.hasName("test-http-server").hasKind(SERVER).hasParent(trace.getSpan(3));
+                }));
+    // cleanup
+    if (channel.get() != null) {
+      channel.get().close();
```
Comment (laurit) [inline instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/client/Netty40ClientSslTest.java:203]: groovy test uses `channel?.close()?.sync()`
channel is also close in a finally block further up, I guess closing it here isn't needed

File: instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/server/Netty40ServerTest.java [hunk 16]
```diff
@@ -0,0 +1,228 @@
...
+
+  @RegisterExtension
+  static final InstrumentationExtension testing = HttpServerInstrumentationExtension.forAgent();
+
+  static final LoggingHandler LOGGING_HANDLER;
+
+  static {
+    InternalLoggerFactory.setDefaultFactory(new Slf4JLoggerFactory());
```
Comment (laurit) [inline instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/server/Netty40ServerTest.java:66]: groovy test does not have this, is it needed?

File: testing-common/src/main/java/io/opentelemetry/instrumentation/testing/junit/http/AbstractHttpClientTest.java [hunk 17]
```diff
@@ -605,6 +605,9 @@ void connectionErrorNonRoutableAddress() {
 
     Throwable thrown =
         catchThrowable(() -> testing.runWithSpan("parent", () -> doRequest(method, uri)));
+    if (thrown == null) {
```
[the remark this answers is not in the document]
Comment (heyams) [reply testing-common/src/main/java/io/opentelemetry/instrumentation/testing/junit/http/AbstractHttpClientTest.java]: removed that check

File: instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/ChannelPipelineTest.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (heyams) [reply instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/ChannelPipelineTest.java:34]: i tried `Class.forName("io.netty.channel.DefaultChannelPipeline")`, but it returned the wrong class though. 
There are a few of this class with the same package name.. 
such as org.asynchttpclient:netty-resolver-dns:2.0.0-RC9
or from different netty-transport ..

is there a way i can force it to retrieve it directly from netty-transport-4.0.0.FInal, which is used in the build.gradle.kts. but when i ran the test in intellij, it already retrieves from the wrong package.

[the remark this answers is not in the document]
Comment (heyams) [reply instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/ChannelPipelineTest.java]: Updated with reflection @laurit

File: instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/ChannelPipelineTest.java (hunk 2 above)
Comment (heyams) [reply instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/ChannelPipelineTest.java]: this is the same though using reflection

File: instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/ChannelPipelineTest.java (hunk 11 above)
Comment (heyams) [reply instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/ChannelPipelineTest.java:133]: updated.. it was out of sync. updated.

File: instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/ChannelPipelineTest.java (hunk 12 above)
Comment (heyams) [reply instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/ChannelPipelineTest.java:167]: updated. out of sync from main branch.

File: instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/ChannelPipelineTest.java [hunk 18]
```diff
@@ -0,0 +1,219 @@
...
+    assertEquals(channelPipeline.first(), httpHandler);
+    assertEquals(channelPipeline.last(), noopHandler);
+    assertEquals(1, channelPipeline.toMap().size());
+
+    ChannelHandler removed = channelPipeline.removeLast();
+    assertEquals(httpHandler, removed);
+    assertEquals(0, channelPipeline.toMap().size());
+
+    // http and instrumentation handlers will be remained; noop handler will be removed
+    // but in reality, line 162 will remove http, instrumentation, and noop handlers
```
Comment (heyams) [inline instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/ChannelPipelineTest.java:173]: please read my comments here.

Comment (laurit) [reply instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/ChannelPipelineTest.java:178]: The issue is in something else. Also note that using line numbers in the comment that way isn't the best idea as the line numbers can easily change.

**Review by laurit — APPROVED** [review]

**Pull request merged**